src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -3847,12 +3841,51 @@ protected Object multiBytes(DynamicObject string, int characterIndex,
 
             // TODO (nirvdrum 22-Dec-16): Since we specialize elsewhere on index being too large, do we need this? Can character boundary search in a CR_BROKEN string cause us to encounter this case?
             if (indexTooLargeProfile.profile(i < k)) {
-                return nil();
+                return -1;
             } else {
                 return p;
             }
         }
 
+        protected boolean characterIndexTooLarge(Rope rope, int characterIndex) {
+            return characterIndex > rope.characterLength();
+        }
+
+        protected boolean characterIndexInBounds(Rope rope, int characterIndex) {
+            return characterIndex >= 0 && !characterIndexTooLarge(rope, characterIndex);
+        }
+
+    }
+
+    // Named 'string_byte_index' in Rubinius.
+    @Primitive(name = "string_byte_index_from_char_index", needsSelf = false, lowerFixnum = 2)
+    @ImportStatic({ StringGuards.class, StringOperations.class })
+    public static abstract class StringByteIndexFromCharIndexNode extends PrimitiveArrayArgumentsNode {
+
+        public static StringByteIndexFromCharIndexNode create() {
+            return StringNodesFactory.StringByteIndexFromCharIndexNodeFactory.create(null);
+        }
+
+        public abstract Object executeFindByteIndex(DynamicObject string, int characterIndex);
+
+        @Specialization(guards = "!characterIndexInBounds(string, characterIndex)")
+        protected Object indexOutOfBounds(DynamicObject string, int characterIndex,
+                @Cached("createBinaryProfile()") ConditionProfile negativeIndexProfile) {
+            if (negativeIndexProfile.profile(characterIndex < 0)) {
+                throw new RaiseException(
+                        getContext().getCoreExceptions().argumentError(
+                                coreStrings().CHARACTER_INDEX_NEGATIVE.getRope(), this));
+            } else {
+                return nil();
+            }
+        }
```

Comment:
Why make this 1 specialization instead of 2?
It's fine, but I think it's just simpler with two specializations, and all guards directly in the guards expressions (no `characterIndexTooLarge`).

Reply:
I just moved this code in this PR. I think the original intent had to with the large number specializations and the annoyance in managing guards. But now that I've split out a lot of the work to `ByteIndexFromCharIndexNode`, that should no longer be a problem.

Reply:
85134a627a3f152abf56aaa1b1fadda9e6e328d4 merged the 2 specializations, no?

Reply:
It looks like you're right. I've worked this over a half dozen times and keep rebasing the commits. This looks like a holdover from a previous attempt and I just got confused on the history.